Card 471 -  Convert api.Config to kapi.List

mfojtik opened the pull request:
This is an initial attempt to move from the Config to upstream List, currently broken ;-)

mfojtik commented:
@smarterclayton this is what I'm currently working on, I'm not sure if the approach is right here. My idea is to support 'Config#v1beta1' in create so we can obsolete the apply. 

I wrote conversion from Config -> List and I updated Template to produce List instead of Config, however I'm not able to get the create work with Config as it is complaining that the Config does not exists in the default version...

smarterclayton commented:
Config in create should work without change.


> On Jan 21, 2015, at 8:12 AM, Michal Fojtik <notifications@github.com> wrote:
> 
> @smarterclayton this is what I'm currently working on, I'm not sure if the approach is right here. My idea is to support 'Config#v1beta1' in create so we can obsolete the apply.
> 
> I wrote conversion from Config -> List and I updated Template to produce List instead of Config, however I'm not able to get the create work with Config as it is complaining that the Config does not exists in the default version...
> 
> —
> Reply to this email directly or view it on GitHub.
>

mfojtik commented:
Config in create never worked

smarterclayton commented:
Flatten() on resource builder should correctly extract "Items"


> On Jan 21, 2015, at 9:41 AM, Michal Fojtik <notifications@github.com> wrote:
> 
> Config in create never worked
> 
> —
> Reply to this email directly or view it on GitHub.
>

mfojtik commented on Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/runtime/helper.go:
```diff
@@ -61,11 +65,16 @@ func ExtractList(obj Object) ([]Object, error) {
-		if !ok {
+		var found bool
+		switch raw.Kind() {
+		case reflect.Interface:
```
@smarterclayton so this works, I did the same to GetItemsPtr just to make sure it works. I also updated the test to exercise 'generic' list as well. Going to submit upstream PR and will update this commit message to include upstream commit ID.

mfojtik commented:
@smarterclayton so I fixed the issue with parsing the Config -> List and the create and get are now working properly. I opened the PR in upstream as well.

Couple of things:

* I deprecated 'apply' by adding `apply` as an alias for `create` and I added deprecation message that saying you should be using `create` command
* I think by this all Apply() code is now deprecated and could be removed (pkg/config)
* I'm working on transforming the Template to `List` as well.

mfojtik commented on pkg/config/api/v1beta1/types.go:
```diff
@@ -5,12 +5,11 @@ import (
-	kapi.TypeMeta   `json:",inline"`
-	kapi.ObjectMeta `json:"metadata,omitempty"`
+	kapi.TypeMeta `json:",inline"`
+	kapi.ListMeta `json:"metadata,omitempty"`
```
fyi, this will remove 'metadata#name' but we were not using 'name' or 'id' for Config anyway (it was optional)

smarterclayton replied:
Yes that's good

mfojtik replied:
it was not :-) I changed it

On Wed, Jan 28, 2015 at 3:41 PM, Clayton Coleman <notifications@github.com>
wrote:

> In pkg/config/api/v1beta1/types.go
> <https://github.com/openshift/origin/pull/698#discussion_r23690563>:
>
> >  type Config struct {
> > -	kapi.TypeMeta   `json:",inline"`
> > -	kapi.ObjectMeta `json:"metadata,omitempty"`
> > +	kapi.TypeMeta `json:",inline"`
> > +	kapi.ListMeta `json:"metadata,omitempty"`
>
> Yes that's good
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/698/files#r23690563>.
>



-- 
-----------------------------------------------
Michal Fojtik <mi@mifo.sk <mfojtik@mifo.sk>>
Red Hat OpenShift, Engineering

mfojtik commented:
@smarterclayton @csrwng this is ready for initial review, please let me know if there is something missing for the https://trello.com/c/t80hvdXJ/471-8-productize-object-list-config-generation-cli-composition

bparees commented on examples/examples_test.go:
```diff
@@ -76,26 +76,8 @@ func validateObject(obj runtime.Object) (errors []error) {
-
 	case *templateapi.Template:
 		errors = templatev.ValidateTemplate(t)
-		for i := range t.Items {
```
isn't this test still valid?

mfojtik replied:
it should test len(errors) > 0 good catch

bparees commented on pkg/cmd/client/kubecfg.go:
```diff
@@ -590,36 +587,8 @@ func (c *KubeConfig) executeTemplateRequest(method string, client *osclient.Clie
-	}
-
-	return true
+	glog.Fatalf("DEPRECATED: THe apply command is now deprecated. Use 'cli create' instead.")
```
since we're symlinking such that you can run "osc" instead, maybe this message should just say "use create instead".

mfojtik commented:
To summarize what this PR does:

* Deprecate `openshift cli apply` (with warning message)
* Nuke `kubecfg`
* Config is now same as api.List (at least internally)
* Nuke all pkg/config code that was related to 'applying a Config' as `openshift cli create` works the same
* Couple small refactorings (template items are now properly validated, label manipulation really belongs to template producing config, not config itself, etc)

smarterclayton commented:
We can't nuke kubecfg until we get resize and run back ported

smarterclayton commented:
That should be in the next rebase, so work with jhon and mark

mfojtik commented:
@smarterclayton resize can be done using https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubectl/cmd/update.go#L68

mfojtik commented:
@smarterclayton I will most likely remove this from this PR to not be blocked on that and then submit another PR when the rebase will be done.

mfojtik commented:
Seems like the kubectl.JSONPrinter is expecting runtime.Object, not the raw JSON we get back from the server after processing. I'm adding a REST client for Template now, so we have an interface for Process...

mfojtik commented:
@smarterclayton can you please review this?

mfojtik commented:
[test]

soltysh commented on pkg/cmd/cli/cli.go:
```diff
@@ -72,6 +74,21 @@ func NewCommandCLI(name string) *cobra.Command {
+// applyToCreate injects the deprecation notice about for 'apply' command into
+// 'create' command.
+// TODO: Remove this once we get rid of 'apply' in all documentation/etc.
+func applyToCreate(dst *cobra.Command) *cobra.Command {
```
How about creating a generic deprecation function that we'll be able to use later?

mfojtik replied:
what other commands we are going to deprecated this way?

soltysh replied:
Hopefully none, but you never can tell :wink:

soltysh commented:
I'd propose leaving at least all the `openshift cli` -> `osc` changes as a separate commit.

mfojtik commented:
@soltysh documentation commit is now a separate commit.

csrwng commented on Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/runtime/helper_test.go:
```diff
@@ -63,6 +63,74 @@ func TestExtractList(t *testing.T) {
+		}
+	}
+}
+
```
It may be useful to add a test case where Items is a pointer and/or interface. Otherwise I don't see how these exercise the new code

mfojtik replied:
should be fixed in upstream

csrwng commented on docs/cli.md:
```diff
@@ -66,14 +66,14 @@ Available formats include:
 
 ### Selectors
 
 cli `get` provides also 'selectors' that you can use to filter the output
```
Shouldn't all occurrences of 'cli' be changed to 'osc'? There's a good number of them i n this file

mfojtik replied:
i'm trying to catch them all, need to run another grep

csrwng commented on docs/cli.md:
```diff
@@ -66,14 +66,14 @@ Available formats include:
 by applying key-value pairs that will be matched with the resource labels:
 
-`$ openshift clic get pods -s template=production`
+`$ oscc get pods -s template=production`
```
oscc (sp?)

mfojtik replied:
s/openshift cli/osc/ ;-) there was a typo before... good catch

csrwng commented on docs/cli.md:
````diff
@@ -156,27 +156,11 @@ cli log
-```
+`$ osc log frontend-pod mysql-container`
 
 cli process
````
osc process ? (as well as other titles on this page)

csrwng commented on pkg/cmd/cli/cli.go:
```diff
@@ -47,11 +48,13 @@ func NewCommandCLI(name string) *cobra.Command {
 	cmds.AddCommand(f.NewCmdDescribe(out))
-	cmds.AddCommand(f.NewCmdCreate(out))
+	cmds.AddCommand(
+		// Deprecate 'openshift cli apply' with 'create' command.
```
'osc apply' ?

csrwng commented on pkg/template/api/validation/validation.go:
```diff
@@ -25,37 +26,16 @@ func ValidateParameter(param *api.Parameter) (errs errors.ValidationErrorList) {
 // ValidateTemplate tests if required fields in the Template are set.
 func ValidateTemplate(template *api.Template) (errs errors.ValidationErrorList) {
-	if len(template.Name) == 0 {
+	if len(template.ObjectMeta.Name) == 0 {
```
ObjectMeta shouldn't be needed here... the way it was before should still work since ObjectMeta is embedded

mfojtik commented:
@csrwng thanks for comments! all addressed

@soltysh fixed 'kubectl' in all commands help as well.

soltysh commented:
@mfojtik thx

mfojtik commented:
@csrwng assets updated, we should see the green colors now

csrwng commented:
LGTM, after a squash

mfojtik commented:
@csrwng you OK with 6 commits? want to keep the upstream separate as well as docs and assets.

mfojtik commented:
@smarterclayton can you please review this before merge?

smarterclayton commented:
I'm adding the osc command to the tar - once that's part of it, and I update the release image, we'll be able to show it.

mfojtik commented:
@smarterclayton I can pull that commit out into a new PR and we can merge it after the 'osc' will be included in the release image.

soltysh commented on pkg/cmd/cli/cmd/process.go:
```diff
@@ -36,78 +56,55 @@ Examples:
+
 	cmd.Flags().StringP("filename", "f", "", "Filename or URL to file to use to update the resource")
 	cmd.Flags().StringP("value", "v", "", "Specify a list of key-value pairs (eg. -v FOO=BAR,BAR=FOO) to set/override parameter values")
 	cmd.Flags().BoolP("parameters", "", false, "Do not process but only print available parameters")
```
I'd parametrized the flag names, since you're using them in the code above. I've tripped myself a couple of times by not doing so :wink:

soltysh commented:
Other than small nit, LGTM :+1:

mfojtik commented:
got 2xLGTM, :punch: [merge]

csrwng commented:
[test]

mfojtik commented:
re [merge]
